src/app/campaign/flight/pingbacks/pingback-form.component.ts

```diff
@@ -0,0 +1,196 @@
+          [errorStateMatcher]="matcher"
+          cdkTextareaAutosize
+          #template="cdkTextareaAutosize"
+          cdkAutosizeMinRows="1"
+          cdkAutosizeMaxRows="5"
+        ></textarea>
+      </mat-form-field>
+      <mat-form-field appearance="outline">
+        <mat-label>Preview</mat-label>
+        <textarea
+          readonly
+          matInput
+          formControlName="preview"
+          [errorStateMatcher]="matcher"
+          cdkTextareaAutosize
+          #preview="cdkTextareaAutosize"
+          cdkAutosizeMinRows="1"
+          cdkAutosizeMaxRows="5"
+        ></textarea>
+      </mat-form-field>
+    </div>
+  `,
+  styleUrls: ['./pingback-form.component.scss'],
+  providers: [
+    { provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(() => PingbackFormComponent), multi: true },
+    { provide: NG_VALIDATORS, useExisting: PingbackFormComponent, multi: true }
+  ],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class PingbackFormComponent implements ControlValueAccessor, OnDestroy {
+  @ViewChild('template', { static: false }) autosizePingback: CdkTextareaAutosize;
+  @ViewChild('preview', { static: false }) autosizePreview: CdkTextareaAutosize;
+  @Input() campaignId: string | number;
+  @Input() flightId: number;
+  @Input() creative: string;
+  @Input() podcastId: string;
+  matcher = new FlightFormErrorStateMatcher();
+  formGroup = this.fb.group({
+    template: ['', [Validators.required, this.validateTemplate.bind(this)]],
+    preview: ['']
+  });
+  valueChangesSub = this.formGroup.get('template').valueChanges.subscribe(template => {
+    this.triggerResize();
+    // set preview or error
+    const previewFormControl = this.formGroup.get('preview');
+    previewFormControl.setValue(this.previewValue(template), { emitEvent: false });
+    previewFormControl.markAsDirty();
+    if (this.formGroup.get('template').getError('error')) {
+      previewFormControl.setErrors({ error: 'Invalid URL' });
+    }
+    if (this.formGroup.valid) {
+      this.onChangeFn(template);
+    }
+  });
+  onChangeFn = (value: any) => {};
+  onTouchedFn = (value: any) => {};
+
+  constructor(private ngZone: NgZone, private fb: FormBuilder) {}
+
+  ngOnDestroy() {
+    if (this.valueChangesSub) {
+      this.valueChangesSub.unsubscribe();
+    }
+  }
+
+  triggerResize() {
+    // Wait for changes to be applied, then trigger textarea resize.
+    this.ngZone.onStable.pipe(take(1)).subscribe(() => {
+      this.autosizePingback.resizeToFitContent(true);
+      this.autosizePreview.resizeToFitContent(true);
+    });
+  }
+
+  previewValue(template: string): string {
+    let preview = '';
+    if (template) {
+      const templateError = this.formGroup.get('template').getError('error');
+      if (templateError) {
+        preview = templateError;
+      } else {
+        preview = template;
+        for (const param of this.findTemplateParams(template)) {
+          if (param === 'campaign' && this.campaignId) {
```

Comment:
TBH, this preview doesn't really make sense to me because we don't have most of these values to replace in the URL template, and when we move it to being in the creative library we literally only have the creative URL. I think we should discuss this validation using a preview on the creative library when talking to ad ops tomorrow. It doesn't seem all that helpful to me.

Reply:
Fair enough.  I'd be fine with just removing the "preview", if the validation message already lists out which/why your url is invalid (checking the params you used).